File: tests/store/store.ts
import * as anchor from "@coral-xyz/anchor";
import { Program, BN, web3 } from "@coral-xyz/anchor";
import { Sodap } from "../../target/types/sodap";
import {
  PublicKey,
  SystemProgram,
  LAMPORTS_PER_SOL,
  Keypair,
  Transaction,
} from "@solana/web3.js";
import { assert } from "chai";
import { describe, it } from "mocha";

type AdminRoleType = { owner: {} } | { manager: {} } | { viewer: {} };
type LoyaltyConfig = {
  pointsPerDollar: BN;
  redemptionRate: BN;
};

describe("Store Tests", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const program = anchor.workspace.Sodap as Program<Sodap>;

  // Test accounts
  const storeOwner = Keypair.generate();
  const storeManager = Keypair.generate();
  const storeViewer = Keypair.generate();

  // PDAs
  let storePda: PublicKey;
  let escrowPda: PublicKey;
  let storeBump: number;
  let escrowBump: number;

  before(async () => {
    // Fund test accounts
    const fundAccounts = [
      storeOwner.publicKey,
      storeManager.publicKey,
      storeViewer.publicKey,
    ];

    for (const account of fundAccounts) {
      const sig = await provider.connection.requestAirdrop(
        account,
        LAMPORTS_PER_SOL
      );
      await provider.connection.confirmTransaction(sig);
    }

    // Find PDAs
    [storePda, storeBump] = PublicKey.findProgramAddressSync(
      [Buffer.from("store"), storeOwner.publicKey.toBuffer()],
      program.programId
    );

    [escrowPda, escrowBump] = PublicKey.findProgramAddressSync(
      [Buffer.from("escrow"), storePda.toBuffer()],
      program.programId
    );
  });

  it("registers a new store with valid data", async () => {
    const storeName = "Test Store";
    const storeDesc = "Basic test store";
    const logoUri = "https://test.com/logo.png";
    const loyaltyConfig = {
      pointsPerDollar: new BN(10),
      redemptionRate: new BN(5),
    };

    // Verify store doesn't exist before registration
    const storeAccountBefore = await provider.connection.getAccountInfo(storePda);
    assert.isNull(storeAccountBefore, "Store should not exist before registration");

    const tx = await program.methods
      .registerStore(
        storeName,
        storeDesc,
        logoUri,
        loyaltyConfig
      )
      .accountsStrict({
        store: storePda,
        authority: storeOwner.publicKey,
        payer: storeOwner.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .signers([storeOwner])
      .rpc();

    // Verify transaction was confirmed
    await provider.connection.confirmTransaction(tx);

    // Fetch and verify store data
    const store = await program.account.store.fetch(storePda);
    assert.equal(store.name, storeName, "Store name mismatch");
    assert.equal(store.description, storeDesc, "Store description mismatch");
    assert.equal(store.logoUri, logoUri, "Store logo URI mismatch");
    assert.equal(store.owner.toString(), storeOwner.publicKey.toString(), "Store owner mismatch");
    assert.isTrue(store.isActive, "Store should be active");
    assert.equal(store.adminCount, 1, "Store should have one admin (owner)");
    assert.equal(store.revenue.toNumber(), 0, "Initial revenue should be 0");
    
    // Verify loyalty config
    assert.equal(
      store.loyaltyConfig.pointsPerDollar.toNumber(),
      loyaltyConfig.pointsPerDollar.toNumber(),
      "Points per dollar mismatch"
    );
    assert.equal(
      store.loyaltyConfig.redemptionRate.toNumber(),
      loyaltyConfig.redemptionRate.toNumber(),
      "Redemption rate mismatch"
    );
  });

  it("updates store information with validation", async () => {
    // Save original store data for comparison
    const originalStore = await program.account.store.fetch(storePda);

    const updatedName = "Updated Store";
    const updatedDesc = "Updated store description";
    const updatedLogoUri = "https://test.com/updated-logo.png";
    const updatedConfig = {
      pointsPerDollar: new BN(20),
      redemptionRate: new BN(10),
    };

    // Try updating with invalid authority (non-owner)
    const nonOwner = Keypair.generate();
    try {
      await program.methods
        .updateStore(
          storePda,
          updatedName,
          updatedDesc,
          updatedLogoUri,
          updatedConfig
        )
        .accountsStrict({
          store: storePda,
          authority: nonOwner.publicKey,
          payer: nonOwner.publicKey,
          systemProgram: SystemProgram.programId,
        })
        .signers([nonOwner])
        .rpc();
      assert.fail("Should not allow non-owner to update store");
    } catch (err: any) {
      assert(err.toString().includes("Unauthorized"));
    }

    // Update with valid owner
    const tx = await program.methods
      .updateStore(
        storePda,
        updatedName,
        updatedDesc,
        updatedLogoUri,
        updatedConfig
      )
      .accountsStrict({
        store: storePda,
        authority: storeOwner.publicKey,
        payer: storeOwner.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .signers([storeOwner])
      .rpc();

    await provider.connection.confirmTransaction(tx);

    // Verify updates
    const updatedStore = await program.account.store.fetch(storePda);
    assert.equal(updatedStore.name, updatedName, "Store name not updated");
    assert.equal(updatedStore.description, updatedDesc, "Store description not updated");
    assert.equal(updatedStore.logoUri, updatedLogoUri, "Store logo URI not updated");
    
    // Verify loyalty config updates
    assert.equal(
      updatedStore.loyaltyConfig.pointsPerDollar.toNumber(),
      updatedConfig.pointsPerDollar.toNumber(),
      "Points per dollar not updated"
    );
    assert.equal(
      updatedStore.loyaltyConfig.redemptionRate.toNumber(),
      updatedConfig.redemptionRate.toNumber(),
      "Redemption rate not updated"
    );

    // Verify unchanged fields remain the same
    assert.equal(
      updatedStore.owner.toString(),
      originalStore.owner.toString(),
      "Store owner should not change"
    );
    assert.equal(
      updatedStore.isActive,
      originalStore.isActive,
      "Store active status should not change"
    );
  });

  it("adds a store admin", async () => {
    const tx = await program.methods
      .addStoreAdmin(storePda, storeManager.publicKey, { manager: {} })
      .accountsStrict({
        store: storePda,
        authority: storeOwner.publicKey,
        payer: storeOwner.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .signers([storeOwner])
      .rpc();

    const store = await program.account.store.fetch(storePda);
    const managerRole = store.adminRoles.find(
      (role) =>
        role.adminPubkey.toString() === storeManager.publicKey.toString()
    );
    assert(managerRole, "Manager role not found");
    assert.deepEqual(managerRole.roleType, { manager: {} });
  });

  it("adds a store viewer", async () => {
    const tx = await program.methods
      .addStoreAdmin(storePda, storeViewer.publicKey, { viewer: {} })
      .accountsStrict({
        store: storePda,
        authority: storeOwner.publicKey,
        payer: storeOwner.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .signers([storeOwner])
      .rpc();

    const store = await program.account.store.fetch(storePda);
    const viewerRole = store.adminRoles.find(
      (role) => role.adminPubkey.toString() === storeViewer.publicKey.toString()
    );
    assert(viewerRole, "Viewer role not found");
    assert.deepEqual(viewerRole.roleType, { viewer: {} });
  });

  it("validates admin role permissions", async () => {
    try {
      await program.methods
        .updateStore(storePda, null, null, null, {
          pointsPerDollar: new BN(10),
          redemptionRate: new BN(5),
        })
        .accountsPartial({
          payer: storeViewer.publicKey,
        })
        .signers([storeViewer])
        .rpc();
      assert.fail("Should not allow viewer to update store");
    } catch (error) {
      const errorMsg = error.toString();
      assert(errorMsg.includes("has_one constraint was violated"));
    }
  });

  it("deactivates and reactivates store", async () => {
    // Deactivate store
    await program.methods
      .updateStore(
        storePda,
        "Test Store",
        "Basic test store",
        "https://test.com/logo.png",
        {
          pointsPerDollar: new BN(10),
          redemptionRate: new BN(5),
        }
      )
      .accountsStrict({
        store: storePda,
        authority: storeOwner.publicKey,
        payer: storeOwner.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .signers([storeOwner])
      .rpc();

    let store = await program.account.store.fetch(storePda);
    assert.isFalse(store.isActive, "Store should be inactive");

    // Reactivate store
    await program.methods
      .updateStore(storePda, null, null, null, {
        pointsPerDollar: new BN(10),
        redemptionRate: new BN(5),
      })
      .accountsStrict({
        store: storePda,
        authority: storeOwner.publicKey,
        payer: storeOwner.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .signers([storeOwner])
      .rpc();

    store = await program.account.store.fetch(storePda);
    assert.isTrue(store.isActive, "Store should be active");
  });

  it("removes a store admin", async () => {
    await program.methods
      .removeStoreAdmin(storePda, storeViewer.publicKey)
      .accountsStrict({
        store: storePda,
        authority: storeOwner.publicKey,
        payer: storeOwner.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .signers([storeOwner])
      .rpc();

    const store = await program.account.store.fetch(storePda);
    const viewerRole = store.adminRoles.find(
      (role) => role.adminPubkey.toString() === storeViewer.publicKey.toString()
    );
    assert.isUndefined(viewerRole, "Viewer role should be removed");
  });

  it("validates loyalty config constraints", async () => {
    // Get the store PDA
    const [storePda] = PublicKey.findProgramAddressSync(
      [Buffer.from("store"), storeOwner.publicKey.toBuffer()],
      program.programId
    );

    // Test points per dollar = 0
    try {
      await program.methods
        .updateStore(storePda, null, null, null, {
          pointsPerDollar: new BN(0),
          redemptionRate: new BN(5),
        })
        .accountsStrict({
          store: storePda,
          authority: storeOwner.publicKey,
          payer: storeOwner.publicKey,
          systemProgram: SystemProgram.programId,
        })
        .signers([storeOwner])
        .rpc();
      assert.fail("Should not allow points per dollar to be 0");
    } catch (error: unknown) {
      const errorMsg = (error as Error).toString();
      assert(errorMsg.includes("Invalid loyalty configuration"));
    }

    // Test redemption rate = 0
    try {
      await program.methods
        .updateStore(storePda, null, null, null, {
          pointsPerDollar: new BN(10),
          redemptionRate: new BN(0),
        })
        .accountsStrict({
          store: storePda,
          authority: storeOwner.publicKey,
          payer: storeOwner.publicKey,
          systemProgram: SystemProgram.programId,
        })
        .signers([storeOwner])
        .rpc();
      assert.fail("Should not allow redemption rate to be 0");
    } catch (error: unknown) {
      const errorMsg = (error as Error).toString();
      assert(errorMsg.includes("Invalid loyalty configuration"));
    }
  });

  it("enforces max admins limit", async () => {
    // Add maximum number of admins
    const maxAdmins = 10; // This should match Store::MAX_ADMINS in the program
    const admins = Array.from({ length: maxAdmins }, () => Keypair.generate());

    // Fund all admin accounts
    for (const admin of admins) {
      await provider.connection.confirmTransaction(
        await provider.connection.requestAirdrop(
          admin.publicKey,
          LAMPORTS_PER_SOL
        )
      );
    }

    // Add admins until we reach the limit
    for (let i = 0; i < maxAdmins - 2; i++) {
      // -2 because owner and manager are already admins
      await program.methods
        .addStoreAdmin(storePda, admins[i].publicKey, { viewer: {} })
        .accountsStrict({
          store: storePda,
          authority: storeOwner.publicKey,
          payer: storeOwner.publicKey,
          systemProgram: SystemProgram.programId,
        })
        .signers([storeOwner])
        .rpc();
    }

    // Try to add one more admin
    try {
      await program.methods
        .addStoreAdmin(storePda, Keypair.generate().publicKey, { viewer: {} })
        .accountsStrict({
          store: storePda,
          authority: storeOwner.publicKey,
          payer: storeOwner.publicKey,
          systemProgram: SystemProgram.programId,
        })
        .signers([storeOwner])
        .rpc();
      assert.fail("Should not allow exceeding max admins");
    } catch (error: unknown) {
      const errorMsg = (error as Error).toString();
      assert(errorMsg.includes("MaxAdminsReached"));
    }
  });
  });
});

File: programs/sodap/src/instructions/store.rs
use crate::error::CustomError;
use crate::state::store::{AdminRole, Store};
use crate::types::{AdminRoleType, LoyaltyConfig};
use crate::Escrow;
use anchor_lang::{prelude::*, Discriminator};

pub fn register_store(
    ctx: Context<RegisterStore>,
    name: String,
    description: String,
    logo_uri: String,
    loyalty_config: LoyaltyConfig,
) -> Result<()> {
    let store = &mut ctx.accounts.store;
    let escrow = &mut ctx.accounts.escrow;

    // Initialize store
    store.owner = ctx.accounts.owner.key();
    store.name = Box::new(name);
    store.description = Box::new(description);
    store.logo_uri = Box::new(logo_uri);
    store.loyalty_config = loyalty_config;
    store.is_active = true;
    store.revenue = 0;
    store.bump = ctx.bumps.store;
    store.escrow_bump = ctx.bumps.escrow;
    store.admin_count = 1;

    // Add owner as first admin
    let admin_role = AdminRole::new(ctx.accounts.owner.key(), AdminRoleType::Owner {});
    store.admin_roles[0] = admin_role;

    // Initialize escrow
    escrow.store = store.key();
    escrow.balance = 0;

    Ok(())
}

#[derive(Accounts)]
#[instruction(
    name: String,
    description: String,
    logo_uri: String,
    loyalty_config: LoyaltyConfig,
)]
pub struct RegisterStore<'info> {
    #[account(
        init,
        payer = payer,
        space = Store::DISCRIMINATOR.len() + Store::LEN,
        seeds = [b"store", owner.key().as_ref()],
        bump
    )]
    pub store: Account<'info, Store>,

    #[account(
        init,
        payer = payer,
        space = Escrow::DISCRIMINATOR.len() + Escrow::LEN,
        seeds = [b"escrow", store.key().as_ref()],
        bump
    )]
    pub escrow: Account<'info, Escrow>,

    pub owner: Signer<'info>,
    #[account(mut)]
    pub payer: Signer<'info>,
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct StoreEscrow<'info> {
    #[account(
        init_if_needed,
        payer = payer,
        space = Escrow::DISCRIMINATOR.len() + Escrow::LEN,
        seeds = [b"escrow", store.key().as_ref()],
        bump
    )]
    pub escrow_account: Account<'info, Escrow>,
    pub store: Account<'info, Store>,
    #[account(mut)]
    pub payer: Signer<'info>,
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
#[instruction(
    name: Option<String>,
    description: Option<String>,
    logo_uri: Option<String>,
    loyalty_config: Option<LoyaltyConfig>,
)]
pub struct UpdateStore<'info> {
    #[account(
        mut,
        seeds = [b"store", owner.key().as_ref()],
        bump = store.bump,
        has_one = owner,
    )]
    pub store: Account<'info, Store>,
    pub owner: Signer<'info>,
    #[account(mut)]
    pub payer: Signer<'info>,
    pub system_program: Program<'info, System>,
}

pub fn update_store(
    ctx: Context<UpdateStore>,
    name: Option<String>,
    description: Option<String>,
    logo_uri: Option<String>,
    loyalty_config: Option<LoyaltyConfig>,
) -> Result<()> {
    let store = &mut ctx.accounts.store;

    if let Some(name) = name {
        store.name = Box::new(name);
    }

    if let Some(description) = description {
        store.description = Box::new(description);
    }

    if let Some(logo_uri) = logo_uri {
        store.logo_uri = Box::new(logo_uri);
    }

    if let Some(loyalty_config) = loyalty_config {
        store.loyalty_config = loyalty_config;
    }

    Ok(())
}

#[derive(Accounts)]
pub struct AddAdmin<'info> {
    #[account(
        mut,
        seeds = [b"store", owner.key().as_ref()],
        bump = store.bump,
        has_one = owner,
    )]
    pub store: Account<'info, Store>,
    pub owner: Signer<'info>,
    #[account(mut)]
    pub payer: Signer<'info>,
    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct RemoveAdmin<'info> {
    #[account(
        mut,
        seeds = [b"store", owner.key().as_ref()],
        bump = store.bump,
        has_one = owner,
    )]
    pub store: Account<'info, Store>,
    pub owner: Signer<'info>,
    #[account(mut)]
    pub payer: Signer<'info>,
    pub system_program: Program<'info, System>,
}

pub fn add_admin(
    ctx: Context<AddAdmin>,
    admin_pubkey: Pubkey,
    role_type: AdminRoleType,
) -> Result<()> {
    let store = &mut ctx.accounts.store;

    // Check if admin already exists
    if store
        .admin_roles
        .iter()
        .any(|role| role.admin_pubkey == admin_pubkey)
    {
        return err!(CustomError::AdminAlreadyExists);
    }

    // Check if max admins reached
    if store.admin_count >= Store::MAX_ADMINS as u8 {
        return err!(CustomError::MaxAdminsReached);
    }

    // Validate role type
    match role_type {
        AdminRoleType::Owner => return err!(CustomError::Unauthorized),
        _ => {}
    }

    // Add new admin
    let admin_count = store.admin_count as usize;
    let admin_role = AdminRole::new(admin_pubkey, role_type);
    store.admin_roles[admin_count] = admin_role;
    store.admin_count += 1;

    Ok(())
}

pub fn remove_admin(ctx: Context<RemoveAdmin>, admin_pubkey: Pubkey) -> Result<()> {
    let store = &mut ctx.accounts.store;
    store.remove_admin(admin_pubkey)?;
    Ok(())
}

pub fn release_escrow(ctx: Context<ReleaseEscrow>, amount: u64) -> Result<()> {
    let cpi_ctx = CpiContext::new(
        ctx.accounts.system_program.to_account_info(),
        anchor_lang::system_program::Transfer {
            from: ctx.accounts.escrow.to_account_info(),
            to: ctx.accounts.owner.to_account_info(),
        },
    );
    anchor_lang::system_program::transfer(cpi_ctx, amount)?;
    ctx.accounts.store.revenue = ctx.accounts.store.revenue.checked_add(amount).unwrap();
    Ok(())
}

pub fn refund_escrow(ctx: Context<RefundEscrow>, amount: u64) -> Result<()> {
    let cpi_ctx = CpiContext::new(
        ctx.accounts.system_program.to_account_info(),
        anchor_lang::system_program::Transfer {
            from: ctx.accounts.escrow.to_account_info(),
            to: ctx.accounts.buyer.to_account_info(),
        },
    );
    anchor_lang::system_program::transfer(cpi_ctx, amount)?;
    Ok(())
}

#[derive(Accounts)]
pub struct ReleaseEscrow<'info> {
    #[account(mut, has_one = owner)]
    pub store: Account<'info, Store>,

    #[account(
        mut,
        seeds = [b"escrow", store.key().as_ref()],
        bump = store.escrow_bump,
    )]
    /// CHECK: escrow vault PDA seeds checked in constraint and bump seed matched in store.escrow_bump
    /// Account data validation not needed as this is just a SOL vault
    pub escrow: AccountInfo<'info>,

    #[account(mut, address = store.owner)]
    pub owner: Signer<'info>,

    pub system_program: Program<'info, System>,
}

#[derive(Accounts)]
pub struct RefundEscrow<'info> {
    #[account(mut)]
    pub store: Account<'info, Store>,

    #[account(
        mut,
        seeds = [b"escrow", store.key().as_ref()],
        bump = store.escrow_bump,
    )]
    /// CHECK: escrow vault PDA seeds checked in constraint and bump seed matched in store.escrow_bump
    /// Account data validation not needed as this is just a SOL vault
    pub escrow: AccountInfo<'info>,

    #[account(mut)]
    pub buyer: Signer<'info>,

    pub system_program: Program<'info, System>,
}
